File: native/engine-chromium-desktop/src/handlers/network_handler.cpp
#include "network_handler.h"

namespace mollotov {

NetworkHandler::NetworkHandler(DesktopHandlerRuntime runtime) : runtime_(std::move(runtime)) {}

void NetworkHandler::Register(DesktopRouter& router) const {
  router.Register("get-network-log",
                  [this](const nlohmann::json& params) { return GetNetworkLog(params); });
  router.Register("clear-network-log",
                  [this](const nlohmann::json&) { return ClearNetworkLog(); });
  router.Register("network-clear", [this](const nlohmann::json&) { return ClearNetworkLog(); });
  router.Register("get-resource-timeline",
                  [](const nlohmann::json&) { return Unsupported("get-resource-timeline"); });
}

nlohmann::json NetworkHandler::GetNetworkLog(const nlohmann::json& params) const {
  if (runtime_.network_store == nullptr) {
    return SuccessResponse({
        {"entries", nlohmann::json::array()},
        {"count", 0},
        {"hasMore", false},
        {"summary", nlohmann::json::object()},
    });
  }

  const auto type_it = params.find("type");
  const auto status_it = params.find("status");
  nlohmann::json entries = ParseJsonText(runtime_.network_store->ToJson());
  const int limit = std::max(1, IntOrDefault(params, "limit", 200));
  nlohmann::json filtered = nlohmann::json::array();

  for (const auto& entry : entries) {
    if (type_it != params.end() && type_it->is_string() &&
        entry.value("category", "") != type_it->get<std::string>()) {
      continue;
    }
    if (status_it != params.end() && status_it->is_string()) {
      const int status_code = entry.value("status_code", 0);
      const std::string filter = status_it->get<std::string>();
      if ((filter == "success" && status_code >= 400) ||
          (filter == "error" && status_code < 400) ||
          (filter == "pending" && status_code != 0)) {
        continue;
      }
    }
    filtered.push_back({
        {"url", entry.value("url", "")},
        {"type", entry.value("category", "Other")},
        {"method", entry.value("method", "GET")},
        {"status", entry.value("status_code", 0)},
        {"statusText", entry.value("status_code", 0) == 0 ? "Pending" : ""},
        {"mimeType", entry.value("content_type", "")},
        {"size", entry.value("size", 0)},
        {"transferSize", entry.value("size", 0)},
        {"timing", {
            {"started", entry.value("start_time", "")},
            {"total", entry.value("duration", 0)},
        }},
        {"initiator", entry.value("initiator", "browser")},
    });
  }

  const std::size_t original_count = filtered.size();
  if (filtered.size() > static_cast<std::size_t>(limit)) {
    filtered.erase(filtered.begin(), filtered.begin() + static_cast<long>(filtered.size() - limit));
  }

  nlohmann::json summary = {
      {"totalRequests", original_count},
      {"totalSize", 0},
      {"totalTransferSize", 0},
      {"byType", nlohmann::json::object()},
      {"errors", 0},
      {"loadTime", 0},
  };
  for (const auto& entry : filtered) {
    summary["totalSize"] = summary["totalSize"].get<int>() + entry.value("size", 0);
    summary["totalTransferSize"] =
        summary["totalTransferSize"].get<int>() + entry.value("transferSize", 0);
    const std::string type = entry.value("type", "Other");
    summary["byType"][type] = summary["byType"].value(type, 0) + 1;
    if (entry.value("status", 0) >= 400) {
      summary["errors"] = summary["errors"].get<int>() + 1;
    }
    summary["loadTime"] = std::max(summary["loadTime"].get<int>(),
                                   entry["timing"].value("total", 0));
  }

  return SuccessResponse({
      {"entries", filtered},
      {"count", filtered.size()},
      {"hasMore", original_count > filtered.size()},
      {"summary", summary},
  });
}

nlohmann::json NetworkHandler::ClearNetworkLog() const {
  if (runtime_.network_store == nullptr) {
    return SuccessResponse({{"cleared", 0}});
  }
  const int count = runtime_.network_store->Count();
  runtime_.network_store->Clear();
  return SuccessResponse({{"cleared", count}});
}

}  // namespace mollotov

File: apps/macos/Kelpie/Handlers/NetworkHandler.swift
import WebKit

/// Handles getNetworkLog and getResourceTimeline.
/// iOS has limited network visibility — uses Performance API (Resource Timing) for basic data.
struct NetworkHandler {
    let context: HandlerContext

    func register(on router: Router) {
        router.register("get-network-log") { body in await getNetworkLog(body) }
        router.register("get-resource-timeline") { body in await getResourceTimeline(body) }
    }

    @MainActor
    private func getNetworkLog(_ body: [String: Any]) async -> [String: Any] {
        let tabId = HandlerContext.tabId(from: body)
        let typeFilter = body["type"] as? String
        let limit = body["limit"] as? Int ?? 200
        let js = """
        (function(){
            var entries = performance.getEntriesByType('resource');
            var nav = performance.getEntriesByType('navigation');
            var all = nav.concat(entries);
            return all.map(function(e){
                var type = 'other';
                if (e.entryType === 'navigation') type = 'document';
                else if (e.initiatorType === 'script') type = 'script';
                else if (e.initiatorType === 'link' || e.initiatorType === 'css') type = 'stylesheet';
                else if (e.initiatorType === 'img') type = 'image';
                else if (e.initiatorType === 'xmlhttprequest') type = 'xhr';
                else if (e.initiatorType === 'fetch') type = 'fetch';
                else if (e.initiatorType === 'font' || (e.name && e.name.match(/\\.(woff2?|ttf|otf|eot)/))) type = 'font';
                return {
                    url: e.name,
                    type: type,
                    method: 'GET',
                    status: e.responseStatus || 200,
                    statusText: 'OK',
                    mimeType: '',
                    size: e.decodedBodySize || 0,
                    transferSize: e.transferSize || 0,
                    timing: {
                        started: new Date(performance.timeOrigin + e.startTime).toISOString(),
                        dnsLookup: Math.round(e.domainLookupEnd - e.domainLookupStart),
                        tcpConnect: Math.round(e.connectEnd - e.connectStart),
                        tlsHandshake: Math.round(e.secureConnectionStart > 0 ? e.connectEnd - e.secureConnectionStart : 0),
                        requestSent: Math.round(e.responseStart - e.requestStart),
                        waiting: Math.round(e.responseStart - e.requestStart),
                        contentDownload: Math.round(e.responseEnd - e.responseStart),
                        total: Math.round(e.duration)
                    },
                    initiator: e.initiatorType || 'other'
                };
            });
        })()
        """
        do {
            let jsonString = try await context.evaluateJSReturningString("JSON.stringify(\(js))", tabId: tabId)
            guard let data = jsonString.data(using: .utf8),
                  let array = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
                return successResponse(["entries": [] as [Any], "count": 0, "hasMore": false, "summary": emptySummary()])
            }
            var filtered = array
            if let typeFilter {
                filtered = filtered.filter { ($0["type"] as? String) == typeFilter }
            }
            let limited = Array(filtered.prefix(limit))
            return successResponse([
                "entries": limited,
                "count": limited.count,
                "hasMore": filtered.count > limit,
                "summary": buildSummary(filtered)
            ])
        } catch {
            if let tabError = tabErrorResponse(from: error) { return tabError }
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func getResourceTimeline(_ body: [String: Any]) async -> [String: Any] {
        let tabId = HandlerContext.tabId(from: body)
        let js = """
        (function(){
            var nav = performance.getEntriesByType('navigation')[0] || {};
            var entries = performance.getEntriesByType('resource');
            return {
                pageUrl: location.href,
                navigationStart: new Date(performance.timeOrigin).toISOString(),
                domContentLoaded: Math.round(nav.domContentLoadedEventEnd || 0),
                domComplete: Math.round(nav.domComplete || 0),
                loadEvent: Math.round(nav.loadEventEnd || 0),
                resources: entries.map(function(e){
                    var type = 'other';
                    if (e.initiatorType === 'script') type = 'script';
                    else if (e.initiatorType === 'link' || e.initiatorType === 'css') type = 'stylesheet';
                    else if (e.initiatorType === 'img') type = 'image';
                    else if (e.initiatorType === 'fetch') type = 'fetch';
                    else if (e.initiatorType === 'xmlhttprequest') type = 'xhr';
                    return {
                        url: e.name,
                        type: type,
                        start: Math.round(e.startTime),
                        end: Math.round(e.startTime + e.duration),
                        status: e.responseStatus || 200
                    };
                })
            };
        })()
        """
        do {
            let result = try await context.evaluateJSReturningJSON(js, tabId: tabId)
            return successResponse(result)
        } catch {
            if let tabError = tabErrorResponse(from: error) { return tabError }
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    private func buildSummary(_ entries: [[String: Any]]) -> [String: Any] {
        var totalSize = 0
        var totalTransfer = 0
        var byType: [String: Int] = [:]
        var errors = 0
        var maxEnd: Double = 0

        for entry in entries {
            totalSize += entry["size"] as? Int ?? 0
            totalTransfer += entry["transferSize"] as? Int ?? 0
            let type = entry["type"] as? String ?? "other"
            byType[type, default: 0] += 1
            let status = entry["status"] as? Int ?? 200
            if status >= 400 { errors += 1 }
            if let timing = entry["timing"] as? [String: Any], let total = timing["total"] as? Int {
                maxEnd = max(maxEnd, Double(total))
            }
        }

        return [
            "totalRequests": entries.count,
            "totalSize": totalSize,
            "totalTransferSize": totalTransfer,
            "byType": byType,
            "errors": errors,
            "loadTime": Int(maxEnd)
        ]
    }

    private func emptySummary() -> [String: Any] {
        ["totalRequests": 0, "totalSize": 0, "totalTransferSize": 0, "byType": [String: Int](), "errors": 0, "loadTime": 0]
    }
}
